Interactive investigation guides form builder UI (#3062)
elastic/security-docs: docs/experimental-features/investigation-guide-actions.asciidoc

[[interactive-investigation-guides]]
== Interactive investigation guides

IMPORTANT: This feature is available for {stack} versions 8.6.0 and newer. Additionally, interactive investigation guides are not compatible between release versions, so query buttons created in 8.6.0 won't render correctly in {version} and vice versa.

Detection rule investigation guides suggest steps for triaging, analyzing, and responding to potential security issues. For custom rules, you can create an interactive investigation guide that includes buttons for launching runtime queries in <<timelines-ui,Timeline>>, using alert data and hard-coded literal values. This allows you to start detailed Timeline investigations directly from an alert using relevant data. 

[role="screenshot"]
image::images/ig-alert-flyout.png[Alert details flyout with interactive investigation guide,550]

> Each query button displays the number of event documents found (in parentheses). Click the button to automatically load the query in Timeline based on settings specified in the investigation guide.

Reviewer: ```suggestion
Each query button displays the number of event documents found in parentheses *( )*. Click the button to automatically load the query in Timeline based on settings specified in the investigation guide.
```
This is a tough one since we usually indicate symbols using parentheses! I'm not sure my suggestion is better than what you had but thought I'd put it out there

Author: Yeah @benironside this sentence has plagued me since last release. 😭 Now I'm thinking of just chucking "in parenthesis" altogether, because the doc count is the only number displayed on the button -- the user will still notice it even if we don't tell them it's in parentheses.
```suggestion
Each query button displays the number of event documents found. Click the button to automatically load the query in Timeline based on settings specified in the investigation guide.
```

You can configure an interactive investigation guide when you <<rules-ui-create,create a new rule>> or <<edit-rules-settings,edit an existing rule>>. 

. When configuring the rule's settings (the *About rule* step for a new rule, or the *About* tab for an existing rule), expand the *Advanced settings*, then scroll down to the *Investigation guide* Markdown editor.
+
[role="screenshot"]
image::images/ig-investigation-guide-editor.png[Investigation guide editor field]

. Place the editor cursor where you want to add the query button in the investigation guide, then select the Insights icon (image:images/ig-insights-icon.png[Insights icon,16,15]) in the toolbar. The *Add investigation query* builder form appears.
+
[role="screenshot"]
image::images/ig-investigation-query-builder.png[Add investigation guide UI,600]

. Complete the query builder form to create an investigation query:
.. *Label*: Enter the text to appear on the query button.
.. *Description*: (Optional) Enter additional text to include with the button.
.. *Filters*: Select fields, operators, and values to build the query. Click *OR* or *AND* to create multiple filters and define their relationships.
+

> To use a value from the alert as a query parameter, enter the field name surrounded by double braces — such as `{{kibana.alert.example}}` — as a custom option for the filter value.

Reviewer: ```suggestion
To use a field value from the alert as a query parameter, enter the field name surrounded by double braces — such as `{{kibana.alert.example}}` — as a custom option for the filter value.
```
Might be slightly more clear what "field name" refers to later in the sentence if it's "field value...field name"

> To use a field value from the alert as a query parameter, enter the field name surrounded by double braces — such as `{{kibana.alert.example}}` — as a custom option for the filter value.

Reviewer 2: QQ: should we call the double brackets "double braces" or "double curly brackets"? I'm using "double curly brackets" [here](https://github.com/elastic/security-docs/pull/3045/files#diff-3ad82a825ca079969d9356d9c29d31c7c3f364c4ca02a12f4a4c2bc28b50ba99R17) but am open to changing it if it should be "double braces".

Author: Good question @nastasha-solomon! From [Wikipedia](https://en.wikipedia.org/wiki/Bracket#Curly_brackets_or_braces) 😵‍💫:

> Curly brackets (British) or braces (American), { and }, are also known as "curly braces",[[30]](https://en.wikipedia.org/wiki/Bracket#cite_note-30) "definite brackets", "swirly brackets", "birdie brackets", "French brackets", "Scottish brackets", "squirrelly brackets",[[31]](https://en.wikipedia.org/wiki/Bracket#cite_note-31) "gullwings", "seagulls", "squiggly brackets", "twirly brackets", "Tuborg brackets" (DK), "accolades" (NL), "pointy brackets", "fancy brackets", "M Braces", "moustache brackets", "squiggly parentheses", or "flower brackets" (India).

I originally went with "double braces" since that's what's in the UI help text, but "double curly brackets" seems more descriptive. Unless we all agree on "squirrelly brackets" instead! 🐿

Reviewer 2: I'm partial to "fancy brackets" 🎩 or "gullwings" 🐦, but "double curly brackets" works too ;)

> To use a field value from the alert as a query parameter, enter the field name surrounded by double braces — such as `{{kibana.alert.example}}` — as a custom option for the filter value.

Author: ```suggestion
To use a field value from the alert as a query parameter, enter the field name surrounded by double curly brackets — such as `{{kibana.alert.example}}` — as a custom option for the filter value.
```

. Save and enable the rule.

[discrete]
[[query-button-syntax]]
==== Query button syntax

The following syntax defines a query button in an interactive investigation guide.

[width="100%",options="header"]
|===
|Field |Description

|`!{investigate{ }}` |The container object holding all the query button's configuration attributes.
|`label` |Identifying text on the button.
|`description` |Additional text included with the button.
|`providers` a|A two-level nested array that defines the query to run in Timeline. Similar to the structure of queries in Timeline, items in the outer level are joined by an `OR` relationship, and items in the inner level are joined by an `AND` relationship.

Each item in `providers` corresponds to a filter created in the query builder UI and is defined by these attributes:

* `field`: The name of the field to query.
* `excluded`: Whether the query result is excluded (such as *is not one of*) or included (*is one of*).
* `queryType`: The query type used to filter events, based on the filter's operator. For example, `phrase` or `range`.

> * `value`: The value to search for. Either a hard-coded literal value, or the name of an alert field (in double braces) whose value you want to use as a query parameter.

Author: ```suggestion
* `value`: The value to search for. Either a hard-coded literal value, or the name of an alert field (in double curly brackets) whose value you want to use as a query parameter.
```
